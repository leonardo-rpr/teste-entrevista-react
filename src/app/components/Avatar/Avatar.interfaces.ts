interface User {
  name: string;
  image: string;
}

export interface AvatarInterface {
  user: User;
  size: string;
}

export interface AvatarStylesInterface {
  image: string;
  size: string;
}
