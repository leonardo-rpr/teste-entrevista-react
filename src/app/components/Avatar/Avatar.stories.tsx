import React from 'react';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import Avatar from './Avatar';

storiesOf('Components', module)
  .addDecorator(withKnobs)
  .add('Avatar', () => (
    <Avatar
      size="50px"
      user={{
        name: text('name', 'Leonardo'),
        image: boolean('with image', false) ? 'https://source.unsplash.com/random?self' : '',
      }}
    />
  ));
