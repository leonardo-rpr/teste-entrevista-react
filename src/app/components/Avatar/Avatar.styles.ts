import styled, { css } from 'styled-components';
import { AvatarStylesInterface } from './Avatar.interfaces';

export const Content = styled.div<AvatarStylesInterface>`
  display: flex;
  align-items: center;
  justify-content: center;
  width: ${({ size }) => size};
  height: ${({ size }) => size};
  border-radius: 100%;
  background-color: #f2295b;
  color: #fff;
  font-size: 40px;

  ${({ image }) => !!image
    && css`
      background: url(${image}) center center no-repeat;
      background-size: cover;
    `}
`;
