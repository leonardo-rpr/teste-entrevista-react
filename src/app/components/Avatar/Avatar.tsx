import React from 'react';
import { AvatarInterface } from './Avatar.interfaces';
import { Content } from './Avatar.styles';

const getInitialLetters = (name: string) => {
  const words = name.trim().split(' ');

  if (words.length > 1) {
    return `${words[0][0]}${words[words.length - 1][0]}`.toUpperCase();
  }

  return name.substring(0, 2).toUpperCase();
};

const Avatar: React.FC<AvatarInterface> = ({ user, size }) => (
  <Content size={size} image={user.image}>
    {!user.image && getInitialLetters(user.name)}
  </Content>
);

Avatar.defaultProps = {
  size: '50px',
};

export default Avatar;
