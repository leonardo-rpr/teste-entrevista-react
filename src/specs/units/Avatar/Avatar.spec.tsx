import React from 'react';
import { render } from '@testing-library/react';
import Avatar from '../../../app/components/Avatar';
import 'jest-styled-components';

describe('Avatar', () => {
  it('Should render correctly without image', () => {
    const user = {
      name: 'Lorem Ipsum',
    };

    const { container } = render(<Avatar user={user} />);
    expect(container.firstChild.textContent).toBe('LI');
    expect(container.firstChild).not.toHaveStyleRule('background-size', 'cover');
    expect(container).toMatchSnapshot();
  });

  it('Should render correctly with single name', () => {
    const user = {
      name: 'Lorem',
    };

    const { container } = render(<Avatar user={user} />);
    expect(container.firstChild.textContent).toBe('LO');
    expect(container).toMatchSnapshot();
  });

  it('Should render correctly with space after name', () => {
    const user = {
      name: 'Lorem ',
    };

    const { container } = render(<Avatar user={user} />);
    expect(container.firstChild.textContent).toBe('LO');
    expect(container).toMatchSnapshot();
  });

  it('Should render correctly with more than two names', () => {
    const user = {
      name: 'Lorem Ipsum dolor',
    };

    const { container } = render(<Avatar user={user} />);
    expect(container.firstChild.textContent).toBe('LD');
    expect(container).toMatchSnapshot();
  });

  it('Should render correctly with image', () => {
    const user = {
      name: 'Lorem Ipsum',
      image: '//img.jpg',
    };

    const { container } = render(<Avatar user={user} />);
    expect(container.firstChild.textContent).toBe('');
    expect(container.firstChild).toHaveStyleRule(
      'background',
      `url(${user.image}) center center no-repeat`,
    );
  });
});
