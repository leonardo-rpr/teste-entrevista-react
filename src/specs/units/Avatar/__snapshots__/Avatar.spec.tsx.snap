// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Should render correctly with more than two names 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 50px;
  height: 50px;
  border-radius: 100%;
  background-color: #f2295b;
  color: #fff;
  font-size: 40px;
}

<div>
  <div
    class="c0"
  >
    LD
  </div>
</div>
`;

exports[`Avatar Should render correctly with single name 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 50px;
  height: 50px;
  border-radius: 100%;
  background-color: #f2295b;
  color: #fff;
  font-size: 40px;
}

<div>
  <div
    class="c0"
  >
    LO
  </div>
</div>
`;

exports[`Avatar Should render correctly with space after name 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 50px;
  height: 50px;
  border-radius: 100%;
  background-color: #f2295b;
  color: #fff;
  font-size: 40px;
}

<div>
  <div
    class="c0"
  >
    LO
  </div>
</div>
`;

exports[`Avatar Should render correctly without image 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 50px;
  height: 50px;
  border-radius: 100%;
  background-color: #f2295b;
  color: #fff;
  font-size: 40px;
}

<div>
  <div
    class="c0"
  >
    LI
  </div>
</div>
`;
